ambry-store/src/main/java/com/github/ambry/store/ScanResults.java

```diff
@@ -26,46 +26,124 @@
+  // subtracted in T0+3H bucket. DELETE[ID2] makes PUT[ID2] invalid, so a delta will be added to T0+2H bucket.
+  //    T0                           T0+1H                   T0+2H                             T0+3H
+  // PUT[ID1]+PUT[ID2]   [-PUT[ID1]      ]    [PUT[ID3]-PUT[ID2] ]    [-PUT[ID1]+PUT[ID1]-PUT[ID3] ]   .....
+  //
+  // This is how delta bucket works. We can pick any point of time and add base validSize and all the delta value in those
+  // buckets prior to this point of time and get the answer we need.
 
+  // Container buckets are very close to the example above, except that it doesn't aggregate all containers' valid data,
+  // rather it keeps them in a map. The key of this map is the AccountId, and the value of this map is yet another map,
+  // whose key is the ContainerId and the value is the valid size of this container.
+  // So {@code containerBaseBucket} is base value for all containers and the {@code containerBuckets} is the delta values
+  // on each time bucket.
+  private final Map<String, Map<String, Long>> containerBaseBucket = new HashMap<>();
+  private final NavigableMap<Long, Map<String, Map<String, Long>>> containerBuckets = new TreeMap<>();
   final long containerForecastStartTimeMs;
   final long containerLastBucketTimeMs;
   final long containerForecastEndTimeMs;
-  final long logSegmentForecastStartTimeMs;
-  final long logSegmentLastBucketTimeMs;
-  final long logSegmentForecastEndTimeMs;
+
+  // LogSegment buckets keep track of valid IndexValue size in each log segments. So the base value of log segment bucket
+  // is a map, whose key is the log segment name and the value is the sum of valid IndexValues' sizes. To test if an
+  // IndexValue is valid or not, we follow the same rule in the {@link BlobStoreCompactor}.
+  //
+  // LogSegment buckets is very close to container buckets, except for several differences.
+  // 1. LogSegment has two delta buckets. one for deleted blobs, the other for expired blobs.
+  // 2. Since we have a retention period for deleted blob, deleted delta's first bucket would start earlier than expired
+  //    delta.
+  // The reason to have two delta buckets is because deleted blobs and expired blobs are invalidated at different time.
+  // When a blob is deleted, we have to wait for a retention duration(7 days in prod) to invalidate it. But when a blob
+  // is expired, it's invalidated right aways. So when calculating valid size for log segment, we have to pass two time
+  // stamps, a delete reference time, usually is now - 7 days, and a expiry time, usually is now.
+  private final NavigableMap<String, Long> logSegmentBaseBucket = new TreeMap<>();
+  private final NavigableMap<Long, NavigableMap<String, Long>> logSegmentBucketsDeltaForDeleted = new TreeMap<>();
+  private final NavigableMap<Long, NavigableMap<String, Long>> logSegmentBucketsDeltaForExpired = new TreeMap<>();
```

Comment:
I think the maps in this class are better to concurrent map as IndexScanner and QueueProcessor may concurrently update the map.

Reply:
QueueProcessor shouldn't be running at the same time with IndexScanner. IndexScanner would change `isScanning` to true while scanning the index files. And when the `isScanning` is true, QueueProcessor won't do anything.